text/0000-custom-test-frameworks.md

```diff
@@ -0,0 +1,376 @@
+
+A profiling framework may want to use this mode to instrument the binary in a certain way. We
+can already do this via proc macros, but having it hook through `cargo test` would be neat.
+
+Overall, it would be good to have a generic framework for post-build steps that can support use
+cases like `#[test]` (both the built-in one and quickcheck), `#[bench]` (both built in and custom
+ones like [criterion]), `examples`, and things like fuzzing. While we may not necessarily rewrite
+the built in test/bench/example infra in terms of the new framework, it should be possible to do so.
+
+The main two features proposed are:
+
+ - An API for crates that generate custom binaries, including
+   introspection into the target crate.
+ - A mechanism for `cargo` integration so that custom test frameworks
+are at the same level of integration as `test` or `bench` as
+   far as build processes are concerned.
+
+ [cargo-fuzz]: https://github.com/rust-fuzz/cargo-fuzz
+ [criterion]: https://github.com/japaric/criterion.rs
+ [Compiletest]: http://github.com/laumann/compiletest-rs
+
+# Detailed proposal
+[detailed-proposal]: #detailed-proposal
+
+(As an eRFC I'm merging the "guide-level/reference-level" split for now; when we have more concrete
+ideas we can figure out how to frame it and then the split will make more sense)
+
+The basic idea is that crates can define test frameworks, which specify 
+how to transform collected test functions and construct a `main()` function,
+and then crates using these can declare them in their Cargo.toml, which will let
+crate developers invoke various test-like steps using the framework.
+
+
+## Procedural macro for a new test framework
+
+A test framework is like a whole-crate procedural
+macro that is evaluated after all other macros in the target crate have
+been evaluated. It is passed the `TokenStream` for every element in the
+target crate that has a set of attributes the test framework has
+registered interest in. For example, to declare a test framework
+called `mytest`:
+
+```rust
+extern crate proc_macro;
+use proc_macro::{TestFrameworkContext, TokenStream};
+
+// attributes() is optional
+#[test_framework]
+pub fn test(context: &TestFrameworkContext) -> TokenStream {
+    // ...
+}
+```
+
+where
+
+```rust
+struct TestFrameworkContext<'a> {
+    items: &'a [AnnotatedItem],
+    // ... (may be added in the future)
+}
+
+struct AnnotatedItem
+    tokens: TokenStream,
+    span: Span,
+    attributes: TokenStream,
+    path: SomeTypeThatRepresentsPathToItem
+}
+```
+
+`items` here contains an `AnnotatedItem` for every item in the
+target crate that has one of the attributes declared in `attributes`
+along with attributes sharing the name of the framework (`test`, here --
+the function must be named either `test` or `bench`).
+
+The annotated function _must_ be named "test" for a test framework and
+"bench" for a bench framework. We currently do not support
+any other kind of framework, but we may in the future.
```

Comment:
What is the significance of the function names? And what is the motivation for this restriction?

Reply:
This is to keep it in line with what other proc macros do -- the function name is the attribute name. Someone asked for it to be restricted to test/bench, I don't remember who. There was a bunch of discussion around not allowing arbitrary "kinds" for now above.